swagger.yaml

```diff
@@ -196,6 +196,44 @@ paths:
                     - $ref: '#/components/schemas/SearchContactsResult'
         '401':
           $ref: '#/components/responses/UnauthorizedError'
+  /cases/search:
+    post:
+      tags:
+        - cases
+      summary: search cases
+      operationId: searchCases
+      parameters:
+        - in: query
+          name: limit
+          required: false
+          schema:
+            type: integer
+            format: int32
+        - in: query
+          name: offset
+          required: false
+          schema:
+            type: integer
+            format: int32
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/SearchContact'
+        description: Case to search
+      responses:
+        '200':
+          description: Search cases result
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  allOf:
+                    - $ref: '#/components/schemas/SequelizeRecord'
+                    - $ref: '#/components/schemas/Case'
```

Comment:
Will we need to return the connected contacts here as well?  What data from the contacts will we need?

Reply:
We'll only need the first connected case. Still thinking if it should come here already, or to fetch it from another end point like `/cases/{caseId}/firstContact`

Reply:
done, created a second endpoint to handle this.

Reply:
Hmm... so I have two thoughts here:
1. In order to render search results after a search (which requires contact details like name), we'll need to make a series of API calls.  Something like:
```
const searchResponse = await post('/cases/search');
const firstContactArray = await Promise.all(searchResponse.forEach( r -> get(`/cases/${r.caseId}/firstContact`)));
```
Separating out `/cases/search` and `/contacts/search` shouldn't have meaningful effects on our latency because we can send them in parallel.  But for this, we have to wait for the first response before the second.  For African helplines, that's two trips across the Atlantic.  While we don't yet have good data to tell us the actual latency of that, I'm not confident the performance hit is worth it here.

2. Something doesn't feel terribly RESTful about `/cases/{caseId}/firstContact`.  It seems over-optimized for a specific use case.  If we want the two-step solution, what about building a more general-use `GET /contacts/{contactId}`, and then in `/cases/search` returning the case's connected contact ids as part of each case, then looking up the contacts with `GET /contacts/{contactId}`?

Reply:
I agree with you that a `/cases/{caseId}/firstContact` seems very specific for a use case. Looking at the options, it seems to me that it's more simple to just return the connected `contacts` on `/cases/search`.

My question to you @nickhurlburt is:
Currently, the other `cases` endpoints don't return the `connectedContacts`. Should we update them to return it, so all `cases` endpoints return a `case` with the same format? (with `connectedContacts` property)

Reply:
@murilovmachado yes, I agree the `cases` endpoints should be consistent.  It makes sense to me to return `connectedContacts`, or at least the ids.  One common REST API trick is to include an `expand` parameter to say which properties should be expanded.  You could imagine an expand parameter with, maybe, three options: `ids`, `first`, `all`.  You always return at least the ids of connected contacts, but can decide whether you want full details.  An example in the wild is this Atlassian documentation: https://developer.atlassian.com/server/confluence/expansions-in-the-rest-api/